### PR TITLE
Version 0.1

### PR DESCRIPTION
**Pull request by AHReccese**
#### Reference Issues/PRs
#1 #4 #5
#### What does this implement/fix? Explain your changes.
Version 0.1 is ready for review.

### CODE REVIEW

**Comment by AHReccese** [comment]
please help me update placeholders for badges in the `README.md` file accordingly.

File: reserver/reserver_obj.py [hunk 1]
```diff
@@ -1,2 +1,107 @@
...
+        generated_built_folder = path.join(getcwd(), "build")
+        generated_dist_folder = path.join(getcwd(), "dist")
+        generated_tar_gz_file = path.join(generated_dist_folder, "*.tar.gz")
+        generated_wheel_file = path.join(generated_dist_folder, "*.whl")
+        # prevent from uploading any other previously build library in this path.
+        if path.exists(generated_dist_folder):
+            rmtree(generated_dist_folder)
+
+        commands = [
+            executable + " -m pip install --upgrade pip",
```
Comment (sepandhaghighi) [inline reserver/reserver_obj.py:63]: It is not recommended to install packages in this way. Instead, we should add them to the `requirements.txt` file.

File: reserver/reserver_obj.py [hunk 2]
```diff
@@ -1,2 +1,107 @@
...
+            executable + " -m pip install setuptools wheel twine",
+            executable + " -m pip install --upgrade setuptools",
+            executable + " " + generated_setup_file_path + " sdist bdist_wheel "]
+        if is_platform_linux():
+            # handle the place build files get generated.
+            # pip install "importlib-metadata<5.0" for
+            # https://bobbyhadz.com/blog/entrypoints-object-has-no-attribute-get
+            commands.append(executable + " -m pip install \"importlib-metadata<5.0\"")
```
Comment (sepandhaghighi) [inline reserver/reserver_obj.py:71]: I am not sure what is happening regarding the error, but the solution seems like a hack (operating system checking)!

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
@AHReccese Thanks for your efforts 💯 
Please take a look at my comments 🔥

File: reserver/reserver_obj.py (hunk 1 above)
Comment (AHReccese) [reply reserver/reserver_obj.py]: Yes, you are right, updated.

File: reserver/reserver_obj.py (hunk 2 above)
Comment (AHReccese) [reply reserver/reserver_obj.py:67]: ok sure, under investigation ...

Comment (AHReccese) [reply reserver/reserver_obj.py]: This issue didn't propagate in GitHub Linux runners, my device had both Python 2 and Python 3 both with multiple versions.
We can ignore this if you want. let me now.
@sepandhaghighi

File: reserver/reserver_func.py [hunk 3]
```diff
@@ -1,2 +1,74 @@
...
+    packages=[],
+    version='0.0.0',
+    description='DESCRIPTION',
+    long_description="LONG_DESCRIPTION",
+    long_description_content_type='text/markdown',
+    author='Development Team',
+    author_email='test@test.com',
+    url='https://url.com',
+    download_url='https://download_url.com',
+    keywords="python3 python",
```
Comment (sepandhaghighi) [inline reserver/reserver_func.py:59]: add this keywords:
`reserved`, `reserve`, `reserver`

File: reserver/reserver_obj.py [hunk 4]
```diff
@@ -1,2 +1,97 @@
...
+        self.password = api_token
+        self.test_pypi = is_test_pypi_account
+
+    def upload_to_pypi(self, package_name):
+        """
+        Upload a template package to pypi or test_pypi.
+
+        :param package_name: package name
+        :type package_name: str
+        :param to_test_pypi: boolean flag to indicate uploading to pypi test or not
```
Comment (sepandhaghighi) [inline reserver/reserver_obj.py:38]: This parameter doesn't exist!

File: reserver/reserver_func.py [hunk 5]
```diff
@@ -1,2 +1,74 @@
...
+    project_urls={
+            'Source': 'https://github.com/source',
+    },
+    install_requires="",
+    python_requires='>=3.6',
+    classifiers=[
```
Comment (sepandhaghighi) [inline reserver/reserver_func.py:65]: Add this classifiers:

`Development Status :: 1 - Planning`
`Programming Language :: Python :: 3.6`
`Programming Language :: Python :: 3.7`
`Programming Language :: Python :: 3.8`
`Programming Language :: Python :: 3.9`
`Programming Language :: Python :: 3.10`
`Programming Language :: Python :: 3.11`
`Programming Language :: Python :: 3.12`

File: reserver/util.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
Comment (sepandhaghighi) [inline reserver/util.py:1]: I suggest moving this function to `reserver_func.py` for better organization and clarity.

File: reserver/reserver_func.py [hunk 7]
```diff
@@ -1,2 +1,74 @@
...
+# invalid email
+# download url
+# url
+# project urls
+
+setup(
+    name =""" + "\"" + package_name + "\"" + """,
+    packages=[],
+    version='0.0.0',
+    description='DESCRIPTION',
```
Comment (sepandhaghighi) [inline reserver/reserver_func.py:52]: This name has been reserved using Reserver

File: reserver/reserver_func.py [hunk 8]
```diff
@@ -1,2 +1,74 @@
...
+    long_description="LONG_DESCRIPTION",
```
Comment (sepandhaghighi) [inline reserver/reserver_func.py:53]: This name has been reserved using [Reserver](https://github.com/openscilab/reserver)

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
@AHReccese Thanks for your efforts 🔥 
Please take a look at my comments 💯

**Comment by AHReccese** [comment]
I added these and now there isn't any other `TODO` left in .md files.
* Reserver Logo
* Add discord badge
* Add pypi badge

**Comment by sadrasabouri** [comment]
Also, I should mention that I used [FlatIcon](https://www.flaticon.com/) for logo components.
@AHReccese @sepandhaghighi Maybe we should add it to the references.

**Comment by sepandhaghighi** [comment]
> Also, I should mention that I used [FlatIcon](https://www.flaticon.com/) for logo components. @AHReccese @sepandhaghighi Maybe we should add it to the references.

@AHReccese Yeah, we should include a mention of it in the `References` section of the `README.md` file

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

File: README.md [hunk 9]
```diff
@@ -97,8 +99,8 @@ Just fill an issue and describe it. We'll check it ASAP! or send an email to [in
  
 You can also join our discord server
 
-<a href="TODO">
-  <img src="TODO" alt="Discord Channel">
+<a href="https://discord.gg/RD2y6SGuY3">
+    <img src="https://img.shields.io/discord/1064533716615049236.svg?style=for-the-badge" alt="Discord Channel">
 </a>
```
Comment (sepandhaghighi) [inline README.md:106]: `## Show your support`

**Comment by AHReccese** [comment]
All requested and found issues resolved
@sepandhaghighi 
ready for final review.

1. Reference part added & indexed(for FlatIcon usage)
2. attached issues resolved
3. docstring & readme updates applied
4. create template package folder added
5. too similar name handled in PyPI main account upload 
6. `.eggfile` "-" issue handled and replaced with "_"

File: README.md [hunk 10]
```diff
@@ -1,18 +1,18 @@
 <div align="center">
-    <img src="https://github.com/openscilab/reserver/raw/main/otherfiles/logo.png" width="280" height="400">
+    <img src="https://github.com/openscilab/reserver/raw/main/otherfiles/logo.png" width="300" height="300">
```
Comment (sepandhaghighi) [inline README.md:2]: logo.png --> reserver.png

Comment (sepandhaghighi) [reply README.md:2]: Alternative: change file name

File: requirements.txt [hunk 11]
```diff
@@ -0,0 +1,3 @@
+setuptools>=40.8.0
```
Comment (sepandhaghighi) [inline requirements.txt:1]: Where is `requests`?!?

File: dev-requirements.txt [hunk 12]
```diff
@@ -0,0 +1,9 @@
+requests>=1.0.0
```
Comment (sepandhaghighi) [inline dev-requirements.txt:1]: Change `>=` to `==` (fix version) for package requirements (`requests`, `setuptools`, `wheel` and `twine`).
To trigger `dependabot` to create a new PR for updating dependencies.

File: reserver/reserver_func.py [hunk 13]
```diff
@@ -1,2 +1,103 @@
 # -*- coding: utf-8 -*-
-"""Functions."""
+"""Reserver Functions."""
```
Comment (sepandhaghighi) [inline reserver/reserver_func.py:2]: Reserver functions.

File: reserver/reserver_obj.py [hunk 14]
```diff
@@ -1,2 +1,100 @@
...
+        :type package_name: str
+        :return: None
+        """
+        if does_package_exist(package_name, self.test_pypi):
+            print("This package already exists in PyPI.")
+            return False 
+        
+        generate_template_setup_py(package_name)
+
+        environ["TWINE_USERNAME"] = self.username
```
Comment (sepandhaghighi) [inline reserver/reserver_obj.py:47]: Please ensure that these two environment variables are deleted or destroyed at the end of their usage.

File: setup.py
Comment (sepandhaghighi) [inline setup.py:1]: Please remove `'Programming Language :: Python :: 3.6',` from this file

File: CHANGELOG.md [hunk 15]
```diff
@@ -5,8 +5,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
-## [0.1] - 2022-08-17
 ### Added
+### Changed
+## [0.1] - 2024-01-31
```
Comment (sepandhaghighi) [inline CHANGELOG.md:10]: Please update this date

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
@AHReccese Thanks for your efforts 🔥 
Please take a look at my comments 💯

**Comment by AHReccese** [comment]
@sepandhaghighi 
Ready for your review.

File: dev-requirements.txt (hunk 12 above)
Comment (sepandhaghighi) [reply dev-requirements.txt]: @AHReccese  Would you please re-check this one? (all basic requirements not only `requests`)

Comment (AHReccese) [reply dev-requirements.txt]: my bad, you are right, hold my beer & lemi apply this.

Comment (AHReccese) [reply dev-requirements.txt]: @sepandhaghighi 
Applied sir.

**Review by sepandhaghighi — APPROVED** [review]
Let's go 🚀

**Pull request merged**